share/picongpu/tests/CurrentDeposition/lib/python/test/CurrentDeposition/grid_class.py

```diff
@@ -0,0 +1,234 @@
+                    s6 = func(end_coord[2] - z)
+
+                    # calculate W and assign it to the grid node
+                    grid_x[z][y][x] = W(s1, s2, s3, s4, s5, s6)
+                    grid_y[z][y][x] = W(s2, s1, s3, s5, s4, s6)
+                    grid_z[z][y][x] = W(s3, s2, s1, s6, s5, s4)
+
+        return grid_x, grid_y, grid_z
+
+    def current_density_field(self, W_grid_x, W_grid_y, W_grid_z, start_coord,
+                              end_coord, Q, cell_depth, cell_width,
+                              cell_height, dt, unit_time, unit_charge,
+                              unit_length, shape):
+        """Calculates the current density in the x, y, and z directions on the
+        grid using the current deposition vector(field) W_grid_x,y,z.
+
+        Parameters:
+        W_grid_x/y/z(np.array) : current deposition vector(field)
+        start_coord(np.array): start_coordinate on minimal grid
+        end_coord(np.array): end_coordinate on minimal grid
+        Q(float): charge of the particle
+        cell_depth/width/height(float): cell dimensions (respectively)
+        dt(float): duration of a timestep
+        unit_time(float): conversion factor of arbitrary PIConGPU time units to
+                          SI
+        unit_charge(float): conversion factor of arbitrary PIConGPU charge
+                            units to SI
+        unit_length(float): conversion factor of arbitrary PIConGPU length
+                            units to SI
+
+
+        Returns:
+        j_grid_x/y/z(np.array): current density vector(field)
+
+        """
+
+        # create null grid to write the current density
+        j_grid_x = np.zeros(shape)
+        j_grid_y = np.zeros(shape)
+        j_grid_z = np.zeros(shape)
+
+        # determine how many grid points are needed to be included
+        # in calculation; aka half-width of assignment function
+        add_active = (self.order + 1) * 1/2
+
+        # determine start x
+        start_i = int(min(start_coord[0], end_coord[0]) - add_active)
+        start_j = int(min(start_coord[1], end_coord[1]) - add_active)
+        start_k = int(min(start_coord[2], end_coord[2]) - add_active)
+
+        # determine end x
+        end_i = int(max(start_coord[0], end_coord[0]) + add_active) + 1
+        end_j = int(max(start_coord[1], end_coord[1]) + add_active) + 1
+        end_k = int(max(start_coord[2], end_coord[2]) + add_active) + 1
+
+        # additive to transform the minimal grid index of the current density
+        # to the equivalent index of the simulation grid.
+        r_i = self.sim_add_x - start_i - 1
+        r_j = self.sim_add_y - start_j - 1
+        r_k = self.sim_add_z - start_k - 1
+
+        combined_factors = (-Q) / dt
+        x_factor = combined_factors * cell_width
+
+        # calculate the norm of the assignment function
+        H = cell_width * cell_height * cell_depth
+        norm = (1/H)**self.order
+
+        for k in range(start_k, end_k, 1):

```

Comment:
I would prefer broadcasting here, e.g.
```python
# assume that start_k already includes the additive constant r_k, etc.
j_grid_x[start_k:end_k, start_j:end_j, start_i:end_i] = x_factor * norm * W_grid_y * j_grid_x[start_k:end_k, start_j-1:end_j-1, start_i:end_i]
```

Reply:
Maybe we just stick with the for-loops, since with negative values or in general start>end due to periodic boundary conditions the slicing does not work anymore.

Reply:
It's fine for now as it resonates with the general style of the code here. But FYI: There's `np.roll` for such occasions.